### ISSUE TITLE
Lifetimes problem with CollisionWorld::interferences_with_ray

### ISSUE THREAD

**Issue by arturoc** [sebcrozet/ncollide#183]
This function: http://ncollide.org/rustdoc/ncollide/world/struct.CollisionWorld.html#method.interferences_with_ray

has the same lifetime for the passed ray and the groups as self which means that something like:

```rust
fn ray_intersect(&self) -> Option<&UserData>{
    let ray = ...;
    self.world.interferences_with_ray(&ray, &self.groups)
}
```

being ray_intersect a function of a struct containing a CollisionWorld and CollisionGroups (although collision groups suffers the same problem if created in the function) won't compile cause ray doesn't live as much as self.

Not sure but wondering if changing the lifetimes of ray and groups to:

```rust
fn interferences_with_ray<'a,'b>(&'a self, ray: &'b Ray<P>, groups: &'b CollisionGroups)
```

would allow to do this, since it doesn't look like the ray or the groups need to live as long as the world.

**Comment by thiolliere** [comment]
(I am on master 0.14 tag)

if you want to be able to write this:
```rust
fn test_ray_intersect() {
    struct Foo {
        world: CollisionWorld<::na::Point2<f32>, ::na::Isometry2<f32>, usize>,
    }

    impl Foo {
        fn ray_intersect(&self) -> Option<(&CollisionObject<::na::Point2<f32>, ::na::Isometry2<f32>, usize>, RayIntersection<::na::Vector2<f32>>)> {
            let ray = unsafe { ::std::mem::uninitialized() };
            let group = unsafe { ::std::mem::uninitialized() };
            self.world.interferences_with_ray(&ray, &group).next()
        }
    }
}
```

then it is possible by adding some lifetimes in the code: diff:
```diff
diff --git a/ncollide_pipeline/world/collision_world.rs b/ncollide_pipeline/world/collision_world.rs
index 5ac086bb..1abf5518 100644
--- a/ncollide_pipeline/world/collision_world.rs
+++ b/ncollide_pipeline/world/collision_world.rs
@@ -260,11 +260,11 @@ impl<P: Point, M: Isometry<P>, T> CollisionWorld<P, M, T> {
 
     /// Computes the interferences between every rigid bodies on this world and a ray.
     #[inline]
-    pub fn interferences_with_ray<'a>(
+    pub fn interferences_with_ray<'a, 'b>(
         &'a self,
-        ray: &'a Ray<P>,
-        groups: &'a CollisionGroups,
-    ) -> InterferencesWithRay<'a, P, M, T> {
+        ray: &'b Ray<P>,
+        groups: &'b CollisionGroups,
+    ) -> InterferencesWithRay<'a, 'b, P, M, T> {
         // FIXME: avoid allocation.
         let mut handles = Vec::new();
         self.broad_phase.interferences_with_ray(ray, &mut handles);
@@ -343,14 +343,14 @@ impl<P: Point, M: Isometry<P>, T> CollisionWorld<P, M, T> {
 }
 
 /// Iterator through all the objects on the world that intersect a specific ray.
-pub struct InterferencesWithRay<'a, P: 'a + Point, M: 'a, T: 'a> {
-    ray: &'a Ray<P>,
+pub struct InterferencesWithRay<'a, 'b, P: 'a + Point, M: 'a, T: 'a> {
+    ray: &'b Ray<P>,
     objects: &'a CollisionObjectSlab<P, M, T>,
-    groups: &'a CollisionGroups,
+    groups: &'b CollisionGroups,
     handles: IntoIter<&'a CollisionObjectHandle>,
 }
 
-impl<'a, P: Point, M: Isometry<P>, T> Iterator for InterferencesWithRay<'a, P, M, T> {
+impl<'a, 'b, P: Point, M: Isometry<P>, T> Iterator for InterferencesWithRay<'a, 'b, P, M, T> {
     type Item = (&'a CollisionObject<P, M, T>, RayIntersection<P::Vector>);
 
     #[inline]
```

same thing should be done then for AABB and point interferences.

Also as a side note Interferences structures are not public and they should be

@sebcrozet if you want I can open 2 PR:
* first with public interferences struct
* second with additional lifetime for the 3 interferences

**Comment by sebcrozet** [comment]
@thiolliere You are correct! I'd be thankful to accept your two PR.

**Comment by sebcrozet** [comment]
This should be fixed by #184. Please reopen if this does not fit your use-case @arturoc .

**Comment by arturoc** [comment]
haven't tried it but that looks correct. thanks!
